New Lattice Universe Update Method

wbinventor opened the pull request:
This is a short PR that introduces a new `Lattice::setUniverses(...)` method. This allows a user to assign a `Universe` to an individual `Lattice` cell as a corollary to assigning all `Universes` in an array to a `Lattice`.

wbinventor commented:
I updated this PR with two new items today:

* Fixed a bug in the deallocation of data from `CPUSolver`
* The "docstring.i" file has been updated from Doxygen

The first item is actually a big deal if one runs 100s-1000s of simulations with different `CPUSolver` instances in the same Python script. Previously the data in a `CPUSolver` wouldn't be deallocated since it did not call the parent `Solver` class' destructor. Now it does so memeory is deallocated as one would expect.

wbinventor commented:
And one more update - a super tiny tweak to fix a memory leak of `LocalCoords` in the `openmoc.plotter` module. This would not have been noticeable, except for cases like mine where you may generate 100s of plots with high pixel resolution. I think this was a bug I actually fixed sometime in the last year, but probably crept back into the code with my PR to refactor `openmoc.plotter` back in December.

wbinventor commented:
Wanna take this one up @geogunow? It's actually relatively short.

samuelshaner commented on openmoc/__init__.py:
```diff
@@ -32,6 +32,7 @@
 time = (now.month, now.day, now.year, now.hour, now.minute, now.second)
 year_string = '-'.join(map(left_pad, map(str, (now.month, now.day, now.year))))
 today_string = ':'.join(map(left_pad, map(str, (now.hour, now.minute, now.second))))
+today_string += ':' + str(now.microsecond)[:2]
```
Have you run into cases where this is needed to avoid having multiple runs write to the same log file? If so, do you think there should be some kind of check to see whether a log file with the desired name already exists?

wbinventor replied:
That's precisely why I added the sub-second resolution to the log filename here.

samuelshaner commented on openmoc/docstring.i:
```diff
@@ -2821,61 +2802,40 @@ the number of threads per block
 ----------
 * res_type :  
-    the type of residuals to compute (SCALAR_FLUX, FISSION_SOURCE, TOTAL_SOURCE)  
+    the type of residual (FLUX, FISSION_SOURCE, TOTAL_SOURCE)  
```
This should still be `SCALAR_FLUX`

wbinventor replied:
Weird...it shows up as `SCALAR_FLUX` in "GPUSolver.cpp" but not here. I'll run "doxy2swig.py" again and see if that fixes things.

wbinventor replied:
Oh I see, it is pulling this comment from "Solver.h" where it was indeed incorrect. Good thing to note for the future - "doxy2swig.py" is smart enough to pull comments from parent classes and inject them as docstrings into their children.

samuelshaner commented on openmoc/docstring.i:
```diff
@@ -2821,61 +2802,40 @@ the number of threads per block
-Creates a polar quadrature object for the GPUSolver on the GPU.  
+Initializes a new PolarQuad object.  
+
+Deletes memory old PolarQuad if one was previously allocated.  
```
Maybe and `of` between `memory` and `old`.

wbinventor replied:
Got it.

samuelshaner commented on openmoc/docstring.i:
```diff
@@ -2966,70 +2911,51 @@ the FSR scalar flux
-\\displaystyle\\sum_{g \\in G} \\nu \\Sigma^F_g \\Phi V_{i}} {\\displaystyle\\sum_{i \\in
-I} \\displaystyle\\sum_{g \\in G} (\\Sigma^T_g \\Phi V_{i} - \\Sigma^S_g \\Phi V_{i} -
-L_{i,g})} $  
+Compute $ k_{eff} $ from total fission and absorption rates in each FSR and energy group.  
```
I think this should now just be `..from the total fission rate in each...`

wbinventor replied:
Actually this comment is wrong - we compute keff from successive fission sources. I'm not sure how this changed in this PR since I never touched it, but anyway.

samuelshaner commented on openmoc/docstring.i:
```diff
@@ -3509,6 +3381,26 @@ Returns
+
+Update the Universe in a particular Lattice cell.  
+
+This method may only be used after an array of Universe has been assigned with the
```
Can `Universe` be made plural or `objects` be added after it?

wbinventor replied:
yep

samuelshaner commented on src/Universe.cpp:
```diff
@@ -1214,6 +1214,41 @@ void Lattice::setUniverses(int num_z, int num_y, int num_x,
 
 /**
+ * @brief Update the Universe in a particular Lattice cell.
+ * @details This method may only be used after an array of Universe
```
I think I should have commented on Universe being plural here instead of in `docstring.i`

wbinventor replied:
No problem, nice to see these things show up in `docstring.i` anyway.

samuelshaner commented:
Glad to see you're pushing the OpenMOC so hard! I think this PR looks good to go as soon as my comments are addressed.

wbinventor commented:
Alright, I updated this based on your comments @samuelshaner. In addition, I added two more things to this PR:

* A new timer report for the `TrackGenerator`
* Moved the loop over x,y coordinates in `openmoc.plotter` into C++ in `Geometry` to dramatically speed up plotting of high-resolution plots

samuelshaner commented on openmoc/numpy_typemaps.i:
```diff
@@ -75,3 +75,9 @@
+
+/* The typemap used to match the method signature for
+ * Geometry::getSpatialDataOnGrid */
+%apply (double* IN_ARRAY1, int DIM1) {(double* gridx, int numx)}
```
Can `numx` be made `num_x` (and same for numy and numxy) here and in `Geometry.cpp` to be consistent with how other methods have `num_x` as inputs?

wbinventor replied:
Done

samuelshaner commented on src/Geometry.cpp:
```diff
@@ -1011,6 +1011,114 @@ void Geometry::computeFissionability(Universe* univ) {
+void Geometry::getSpatialDataOnGrid(double* gridx, int numx,
+				    double* gridy, int numy,
+				    int* domains, int numxy,
+				    double zcoord, const char* domain_type) {
```
We do have the ability to make plots in the xz and yz planes with the plotter. Is that functionality preserved? It looks like this is specific to making xy plots.

wbinventor replied:
As we've determined, 3D plotting is not currently supported by `openmoc.plotter` and it is not entirely trivial to add in.

samuelshaner commented on src/TrackGenerator.cpp:
```diff
@@ -2036,3 +2049,38 @@ double TrackGenerator::getPhi(int azim) {
+ *        code in the loop over track for ray tracing.
+ */
+void TrackGenerator::clearTimerSplits() {
+  _timer->clearSplit("Total time");
```
Could this be `_timer->clearSplits()` in case further timer splits are added?

wbinventor replied:
Sure, changed it here and in `Solver`

wbinventor commented:
I believe this PR is ready for final review / merging. Do note that I made a few updates since the last review:

* performance optimization for `openmoc.process.get_scalar_fluxes(...)` method when retrieving fluxes for *all* FSRs and energy groups
* 4-space indentation in `thisown.i` since the previous 2-space indentation breaks some newer versions of SWIG shipped with Anaconda
* bug fix for `openmoc.process.Mesh.get_mesh_cell_indices(...)` to work for `Meshes` which cover only a subset of the entire `Geometry`
* small change to the way the `plots` directory is created in `openmoc.plotter` to eliminate race conditions if multiple MPI processes are in use with `mpi4py`

@samuelshaner @geogunow can we expedite this merge in the next day or two? I'm trying to sync @logty with a lot of my stuff including this code, and it helps to have everything on the "develop" branches for each repo if possible.

geogunow commented on openmoc/numpy_typemaps.i:
```diff
@@ -75,3 +75,9 @@
+ * Geometry::getSpatialDataOnGrid */
+%apply (double* IN_ARRAY1, int DIM1) {(double* grid_x, int num_x)}
+%apply (double* IN_ARRAY1, int DIM1) {(double* grid_y, int num_y)}
+%apply (int* ARGOUT_ARRAY1, int DIM1) {(int* domains, int num_xy)}
```
Why not just use ``std::vector`` ?

wbinventor replied:
I'm open to using `std::vector`, but since I don't know how to, I went with NumPy typemaps. Do we have any examples which use `std::vector`? Is this documented in our Developer's Guide as are the NumPy typemaps? If not, I would opt to leave this as is for now in lieu of an overhaul of our NumPy typemaps to Python lists / `std::vectors` for this method as well as many others in the future.

geogunow commented on openmoc/thisown.i:
```diff
@@ -107,6 +107,23 @@
+%pythonappend Lattice::updateUniverse %{
+  # SWIG 3
+  if 'universe' in locals():
+    universe = locals()['universe']
```
This confuses me. From my interpretation of what ``locals()`` does, this line should do nothing

wbinventor replied:
This line should extract the `universe` variable from the list of local variables. Basically, it's a way of getting access to the parameter called `locals` for the `Lattice::updateUniverse(...)` method.

geogunow commented on src/Geometry.cpp:
```diff
@@ -1011,6 +1011,114 @@ void Geometry::computeFissionability(Universe* univ) {
+void Geometry::getSpatialDataOnGrid(double* grid_x, int num_x,
+				    double* grid_y, int num_y,
+				    int* domains, int num_xy,
+				    double zcoord, const char* domain_type) {
```
So yeah, this could be made much more compact with ``std::vector`` and easier to translate with Python lists

wbinventor replied:
It would be great to minimize this code since I do recognize this is quite clunky. I'd just rather not go hunting on another SWIG expedition to learn how to use `std::vector`. But if we have some examples in place already then I'll build upon them.

geogunow replied:
If you use std::vector no typemap is needed. You just need to include
std_vector.i  (which we already do) and it will just work.

On Mon, Apr 4, 2016, 9:03 PM Will Boyd <notifications@github.com> wrote:

> In src/Geometry.cpp
> <https://github.com/mit-crpg/OpenMOC/pull/293#discussion_r58473658>:
>
> > + *              grid_x, grid_y, num_xy, zcoord, domain_type)
> > + * @endcode
> > + *
> > + * @param grid_x a NumPy array of the x-coordinates
> > + * @param num_x the number of x-coordinates in the grid
> > + * @param grid_y a NumPy array of the y-coordinates
> > + * @param num_y the number of y-coordinates in the grid
> > + * @param domains a NumPy array of the domain IDs
> > + * @param num_xy the number of coordinates in the 2D grid
> > + * @param zcoord the z-coordinate to use to find the domain IDs
> > + * @param domain_type the type of domain ('fsr', 'material', 'cell')
> > + */
> > +void Geometry::getSpatialDataOnGrid(double* grid_x, int num_x,
> > +				    double* grid_y, int num_y,
> > +				    int* domains, int num_xy,
> > +				    double zcoord, const char* domain_type) {
>
> It would be great to minimize this code since I do recognize this is quite
> clunky. I'd just rather not go hunting on another SWIG expedition to learn
> how to use std::vector. But if we have some examples in place already
> then I'll build upon them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/293/files/e705c7159561da3597643f2d7a18d5ce67bbf8a4#r58473658>
>

wbinventor replied:
Ok then, so what would this method signature look like in that case? And how would one call it from Python?

geogunow replied:
Just like you would in C++. SWIG can automatically translate lists/numpy
arrays to vectors and vice versa.

On Mon, Apr 4, 2016, 9:06 PM Will Boyd <notifications@github.com> wrote:

> In src/Geometry.cpp
> <https://github.com/mit-crpg/OpenMOC/pull/293#discussion_r58473925>:
>
> > + *              grid_x, grid_y, num_xy, zcoord, domain_type)
> > + * @endcode
> > + *
> > + * @param grid_x a NumPy array of the x-coordinates
> > + * @param num_x the number of x-coordinates in the grid
> > + * @param grid_y a NumPy array of the y-coordinates
> > + * @param num_y the number of y-coordinates in the grid
> > + * @param domains a NumPy array of the domain IDs
> > + * @param num_xy the number of coordinates in the 2D grid
> > + * @param zcoord the z-coordinate to use to find the domain IDs
> > + * @param domain_type the type of domain ('fsr', 'material', 'cell')
> > + */
> > +void Geometry::getSpatialDataOnGrid(double* grid_x, int num_x,
> > +				    double* grid_y, int num_y,
> > +				    int* domains, int num_xy,
> > +				    double zcoord, const char* domain_type) {
>
> Ok then, so what would this method signature look like in that case? And
> how would one call it from Python?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/293/files/e705c7159561da3597643f2d7a18d5ce67bbf8a4#r58473925>
>

wbinventor replied:
Wow, the `std_vector.i` typemap is fantastic! It was such an easy drop-in replacement, and didn't require nearly the painful wrangling that the NumPy typemaps often incur. We should absolutely start using `std_vector.i` in place of NumPy typemaps everywhere, if we can define our own `UniverseVector`, `CellVector`, `MaterialVector`, etc.

Also @geogunow just to follow up on our conversation from yesterday, the input args and return value are indeed NumPy arrays. Though I presume the typemap may also be compatible with lists.

geogunow commented on src/Material.cpp:
```diff
@@ -563,7 +563,7 @@ void Material::setSigmaT(double* xs, int num_groups) {
     /* If the cross-section is near zero (e.g., within (-1E-10, 1E-10)) */
     if (fabs(xs[i]) < ZERO_SIGMA_T) {
-      log_printf(INFO, "Overriding zero cross-section in "
+      log_printf(INFO, "Overriding zero tot. MGXS in "
```
Why "tot." instead of "total"?

wbinventor replied:
Yeah so the reason is simply that the message length with "total" exceeds 80 characters and is split across two lines. This can clutter an output log when you have lots (100s-1000s) of `Materials` with lots (10s) of energy groups. If you think "tot." is confusing then I'll switch it to "total" though.

geogunow commented on src/Material.cpp:
```diff
@@ -586,7 +586,7 @@ void Material::setSigmaTByGroup(double xs, int group) {
     /* If the cross-section is near zero (e.g., within (-1E-10, 1E-10)) */
     if (fabs(xs) < ZERO_SIGMA_T) {
-      log_printf(INFO, "Overriding zero cross-section in "
+      log_printf(INFO, "Overriding zero tot. MGXS in "
```
ditto

samuelshaner commented on src/TrackGenerator.cpp:
```diff
@@ -2036,3 +2049,38 @@ double TrackGenerator::getPhi(int azim) {
+  /* Get the total runtime */
+  double tot_time = _timer->getSplit("Total time");
+  msg_string = "Total time for ray tracing";
+  msg_string.resize(53, '.');
```
What do you think about having 53 be a constant, like `REPORT_WIDTH`? I'm putting some reports for memory and run parameters in ClosedMOC and am finding that I'm using this value a lot and a constant might be a little cleaner to use throughout.

wbinventor replied:
Sure, I just added a new `REPORT_WIDTH` macro to "constants.h" and dropped it into the timer reports in `Solver` and `TrackGenerator'. Perhaps a formalized "report" interface in the `log` module would be useful (and minimize code reuse) for these new timer / memory footprint reports you speak of @samuelshaner.

wbinventor commented:
@geogunow @samuelshaner this up-to-date with respect to your comments. I'm going to test out the `GPUSolver` and `VectorizedSolver` to ensure that they are still in working order, but otherwise believe this is ready to go.

wbinventor commented:
The `GPUSolver` looks good but the `VectorizedSolver` is not test-able for an unrelated compilation issue recorded in #300.

wbinventor commented:
Just a heads up that I tacked on another minor feature in this PR - a new optional `store` parameter to `TrackGenerator::generateTracks()` which can allow one to toggle whether or not to store tracks to a file. I've found that in some cases when multiple OpenMOC processes (*e.g.*, jobs running concurrently on a cluster) attempt to store their tracks to the same file at the same time it may lead to `std::bad_alloc` errors. In these cases, it is beneficial to just dumping the tracks to a file altogether.

wbinventor commented:
Just a heads up that I updated the SPH factor sample inputs to be compatible with:

1) the latest version of the `openmc.mgxs` module (which added a new attribute to the `MGXS`)
2) the reformatted `openmoc.opencg_compatible` module from #292.
